<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="David Gerard" />

<meta name="date" content="2019-09-09" />

<title>Parallelized Genotyping using Updog</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Parallelized Genotyping using Updog</h1>
<h4 class="author">David Gerard</h4>
<h4 class="date">2019-09-09</h4>



<div id="abstract" class="section level1">
<h1>Abstract</h1>
<p>We provide two examples of parallelizing <code>flexdog</code> to genotype many genes.</p>
</div>
<div id="analysis" class="section level1">
<h1>Analysis</h1>
<p>Before we begin, we’ll need to load the <code>parallel</code>, <code>foreach</code>, and <code>doParallel</code> packages. There are other ways to perform parallel computation in R, but we find these to be pretty handy and easy to use.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(parallel)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">library</span>(foreach)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">library</span>(doParallel)</a></code></pre></div>
<p>Of course, we’ll need to load the <code>updog</code> package to actually do the genotyping.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">library</span>(updog)</a></code></pre></div>
<p>We’ll work with the <code>uitedewilligen</code> dataset, which is a subset of the data from <span class="citation">Uitdewilligen (2013)</span>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">data</span>(<span class="st">&quot;uitdewilligen&quot;</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">ngenes &lt;-<span class="st"> </span><span class="kw">ncol</span>(uitdewilligen<span class="op">$</span>refmat)</a></code></pre></div>
<p>You can read more about these data by typing <code>?uitdewilligen</code> in R.</p>
<p>Suppose we wanted to use parallel computing to genotype this individual. Also suppose we only wanted the MAP (maximum a posteriori) estimated genotypes for each individual and each gene. Then we can use the following code.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">nc &lt;-<span class="st"> </span><span class="dv">2</span> <span class="co">## number of cores. </span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">        <span class="co">## You should change this for your specific computing environment.</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">cl &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw">makeCluster</span>(nc)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">doParallel<span class="op">::</span><span class="kw">registerDoParallel</span>(<span class="dt">cl =</span> cl)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="kw">stopifnot</span>(foreach<span class="op">::</span><span class="kw">getDoParWorkers</span>() <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) <span class="co">## make sure cluster is set up.</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6">gene_est &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="dv">1</span><span class="op">:</span>ngenes, </a>
<a class="sourceLine" id="cb4-7" data-line-number="7">                    <span class="dt">.combine =</span> cbind, </a>
<a class="sourceLine" id="cb4-8" data-line-number="8">                    <span class="dt">.export =</span> <span class="st">&quot;flexdog&quot;</span>) <span class="op">%dopar%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">                      <span class="co">## fit flexdog</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10">                      fout &lt;-<span class="st"> </span><span class="kw">flexdog</span>(<span class="dt">refvec  =</span> uitdewilligen<span class="op">$</span>refmat[, i], </a>
<a class="sourceLine" id="cb4-11" data-line-number="11">                                      <span class="dt">sizevec =</span> uitdewilligen<span class="op">$</span>sizemat[, i], </a>
<a class="sourceLine" id="cb4-12" data-line-number="12">                                      <span class="dt">ploidy  =</span> uitdewilligen<span class="op">$</span>ploidy, </a>
<a class="sourceLine" id="cb4-13" data-line-number="13">                                      <span class="dt">model   =</span> <span class="st">&quot;norm&quot;</span>, </a>
<a class="sourceLine" id="cb4-14" data-line-number="14">                                      <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb4-15" data-line-number="15">                      fout<span class="op">$</span>geno</a>
<a class="sourceLine" id="cb4-16" data-line-number="16">                      }</a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="kw">stopCluster</span>(cl)</a></code></pre></div>
<p>Note that I used <code>model = &quot;norm&quot;</code> <span class="citation">(Gerard and Ferrão 2019)</span> because we have a very small number of samples.</p>
<p>However, this doesn’t return any sort of diagnostic statistics for each SNP. So suppose we now want to save all of the output of each SNP and not just the MAP genotype, then we can use the following code:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">nc &lt;-<span class="st"> </span><span class="dv">2</span> <span class="co">## number of cores. </span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">        <span class="co">## You should change this for your specific computing environment.</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3">cl &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw">makeCluster</span>(nc)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">doParallel<span class="op">::</span><span class="kw">registerDoParallel</span>(<span class="dt">cl =</span> cl)</a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw">stopifnot</span>(foreach<span class="op">::</span><span class="kw">getDoParWorkers</span>() <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) <span class="co">## make sure cluster is set up.</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6">glist &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="dv">1</span><span class="op">:</span>ngenes, </a>
<a class="sourceLine" id="cb5-7" data-line-number="7">                 <span class="dt">.export =</span> <span class="st">&quot;flexdog&quot;</span>) <span class="op">%dopar%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb5-8" data-line-number="8">                   <span class="co">## fit flexdog</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9">                   fout &lt;-<span class="st"> </span><span class="kw">flexdog</span>(<span class="dt">refvec  =</span> uitdewilligen<span class="op">$</span>refmat[, i], </a>
<a class="sourceLine" id="cb5-10" data-line-number="10">                                   <span class="dt">sizevec =</span> uitdewilligen<span class="op">$</span>sizemat[, i], </a>
<a class="sourceLine" id="cb5-11" data-line-number="11">                                   <span class="dt">ploidy  =</span> uitdewilligen<span class="op">$</span>ploidy, </a>
<a class="sourceLine" id="cb5-12" data-line-number="12">                                   <span class="dt">model   =</span> <span class="st">&quot;norm&quot;</span>, </a>
<a class="sourceLine" id="cb5-13" data-line-number="13">                                   <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-14" data-line-number="14">                   fout</a>
<a class="sourceLine" id="cb5-15" data-line-number="15">                   }</a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="kw">stopCluster</span>(cl)</a></code></pre></div>
<p>Returning the diagnostics for each SNP is pretty easy. For example, below we extract the variable <code>prop_mis</code> from the list <code>glist</code> and make a histogram of these values.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">prop_mis &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="dt">X =</span> glist, <span class="dt">FUN =</span> <span class="cf">function</span>(x) x<span class="op">$</span>prop_mis)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="kw">qplot</span>(prop_mis, </a>
<a class="sourceLine" id="cb6-4" data-line-number="4">      <span class="dt">xlab =</span> <span class="st">&quot;Posterior Proportion Mis-genotyped&quot;</span>, </a>
<a class="sourceLine" id="cb6-5" data-line-number="5">      <span class="dt">bins =</span> <span class="dv">30</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="st">  </span><span class="kw">theme_bw</span>()</a></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAbAAAAFQCAMAAADp1NuaAAACvlBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYoKCgpKSkqKiosLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFERERFRUVHR0dISEhJSUlLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVXV1dYWFhZWVlaWlpbW1tcXFxeXl5fX19hYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx+fn5/f3+AgICCgoKDg4OFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKUlJSVlZWWlpaXl5eYmJiZmZmcnJydnZ2enp6fn5+goKChoaGkpKSlpaWmpqanp6eoqKirq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///+oh2EqAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAOdklEQVR4nO3d/38UxR3H8UVikRarFTRWS/1W+s3WRmvVopTaNkqligSkFEGsokL8WlQsBdtSAYGY2C+W+t3aGr5DJMFgW40W0ChiUCB3uW+5y9e7+S+6e3t7N7d3Ozf7mUvu9vp+/QDJ7szsek+zBB/nRGPIU2mlvgHkLoB5LIB5LIB5LIB5LIB5LJdgO+7NV/099XmPS7ZUZfKye1Rm1y9Tmb1U7c5dvWhP08Ca9uc7Otg97G6ZrBIhhcks6leZPdCnMjsQVpnd6+ZFG74r9QHAFAIYIYAJAxgfwAgBTBjA+ABGCGDCAMYHMEIAEwYwPoARApgwgPF5EuzmTO6WSgcwiQBmBjBCAJMIYGYAIwQwiahgDS29tjgw+ynJIn7ixGRBn8rsUEhltr9HZXYg4mJw6E4aWOO+IVscmP2UZINB4sRkEb/K7FhUZXZPSGV2eMDF4P4lNDA8Evk88EgEGB/ACAFMGMD4AEYIYMIAxgcwQgATBjA+gBECmDCA8QGMEMCEAYwPYIQAJgxgfAAjBDBhAOMDGCGACQMYH8AIAUwYwPgARghgwgDGBzBCAON7Y6H+y78Wz11vrQ0wvrIDOz7vVsZisw8NLH8hdQRgfOUGNlTfrIPtfpixN+tThwDGV25gT778kQ723JOMHf25/qlvzpw5d+4M2OLA7Kdk81EnGvm7VWb3+FVmd6vdeY+Lwb47CoPtWsWSYBsY65qnfx5asWLFg3sitjgw+ynZ/NSJRgGfyuxQUGW2r0dldiDsYnBQ4n+GWHlT3azr6oK7HmHs38tSx/BI5Cu3RyJLfoVF53QlVj6fOgAwvvIEY/tvX7gO39bnq/zAcgIYH8AIAUwYwPgARghgwgDGBzBCABMGMD6AEQKYMIDxAYwQwIQBjA9ghAAmDGB8ACMEMGEA4wMYIYAJAxgfwAgBTBjA+ABGCGDCAMYHMEIAEwYwPg+APdXab4sDs5+SrC9AnJgs7FOZ3RtRme1XuvNQzMXgKHHfeoDxeQAMj0Q+DzwSAcYHMEIAEwYwPoARApgwgPEBjBDAhAGMD2CEACYMYHwAIwQwYQDjAxghgAkDGB/ACAFMGMD4AEYIYMIAxgcwQgATBjA+gBECmDCA8QGMEMCEAYyv3MCeuaXusTB+0IBzZQb21rzu/sca8IMGnCszsCNvsXhjY+YHDQTu1dsdtsWB2U/J5qdONAr4VGYHAyqzfUp33hNyMTggsW89Y1tnLOrDDxpwrsx+0IBez9o1mR80YIRHIl+ZPRK37GCscwHDDxpwrMzAdi2OJhpX4QcNOFdmYImGubesDOIHDThXZmD5AhgfwAgBTBjA+ABGCGDCAMYHMEIAEwYwPoARApgwgPEBjBDAhAGMD2CEACYMYHwAIwQwYQDjAxghgAkDGB/ACAFMGMD4AEYIYMIAxgcwQgAT1rhvyBYHZj8l2WCQODFZxK8yOxZVmd0TUpkdHnAxuH8JDayhpdcWB2Y/JVnET5yYLOhTmR0Kqcz296jMDkRcDA5h3/pkFf9IBBgfwAgBTBjA+ABGCGDCAMYHMEIAEwYwPoARApgwgPEBjBDAhAGMD2CEACYMYHwAIwQwYQDjAxghgAkDGB/ACAFMGMD4AEYIYMIAxgcwQgATBjC+cgPbsqBu+THsW+9cmYEdnfnJQMMD2LfeuTID27+esY/nZvat721qanr09ZgtDsx+SrJogDgxWcinMjsSVpntV7rzYK+LwRG5t2oPrd6Y2bf++LRp0xZu99niwOynZOumTiz17G612W4GH/+lDFjLoqY49q13rsweiYkn7j+q/4Z96x0rM7C2u4fi8Tj2rXeuzMAaa/XqsG+9c2UGli+A8Xkd7GYaHsAkApgZwAAmHcAIAQxg0gGMEMAAJh3ACAEMYNIBjBDAACYdwAgBDGDSAYwQwAAmHcAIAQxg0gGMEMAAJh3ACAEMYNKNzr717QlbDmD2YYLiIReDc4r6VWb3x1RmB8Iqs3uHXAweIoI17A7bcgCzDxMU8rsYnFPApzI7GFCZ7VO6856Qi8GBO2hgeCTyeeCRCDA+gBECmDCA8QGMEMCEAYwPYIQAJgxgfAAjBDBhAOMDGCGACQMYH8AIAUwYwPgARghgwgDGBzBCABMGMD6AEQKYMIDxAYwQwIQBjA9ghAAmDGB8ZQc2vCDKXOxbDzBXFR/s1WW1Opj8vvUAc1Xxwd5qnaGDpfet72tubl7d2m/LAcw+TFBfwMXgnMI+ldm9EZXZfqU7D8VcDI7K7Vs/UwdL71v/aU1Nzfxt3bYcwOzDkFrHpPatN8GsfeuHu7q6NrwxbMsBzD5M0FDQxeCcev0qs/uiKrN7QiqzwwMuBg8skQaT37cef4a5aiS+rTfA5PetB5irRgpMft96gLkKf3EmBDCASQcwQgADmHQAIwQwgEkHMEIAA5h0ACMEMIBJBzBCAAOYdOUF5sIOYBIBzAxgAJMOYIQABjDpAEYIYACTDmCEAAYw6QBGCGAAk25UwKT3recqtCU79q2XaMT3rcdXWMHwSCQEMIBJBzBCAAOYdJ4Ay3cYYBIBzAxgAJMOYIQABjDpAEYIYPJgDiPcpboIwAAmHcAIAQxg0gGMEMCUc1ja4X4KL+IwOnXEPRh/GYABzCmAmQEMYNIBDGC5AQxgcjncT+HR4hFJsMJXdFik2GCCf2CAmVUkGGnf+mLncG+FR4tHVCIYbd/6Yudwc4VHi0dUIlh633pXu2oXO4fNpguPFo9I7qpd+IoOixR7V23BP7DkrtpGtH3ri53Ddu6FR0usRx1R9M35BWtL7lufBLP2rR9oa2v7fdtgnmLd/fkOSzYQVJg8GPGrzI71qszuUbrzsJsXrU/+K6zAvvVGg91u/kphD28RkMjFn2EF9q03AhilEfuLs3jfeiOAUSrRf+kwAhglgBECGCGASQQwM4ARAphEADMDGCGASUQG+2Nbnlpe25vvsGSt2xUmt+3aojK7ZbfK7K1Kd77TzYv2OhHs7aZ8rZm/Lu9xyRpUJi9frDL7qadUZi+5X2V2Q6Ob0VtpYPnbV9NZjGVIrfthyS7N5j00+tcEmEIAIwQwQv1dg8VYhlTok5Jdmp1Q+g6VVlHA0OgFMI8FMI+lAJZ5v2LqI/4NjCNczrW3LKhbfqxE12bsjYWjc2mmApZ5v2Lqo6w3MI5sOdc+OvOTgYYHSnNtxo7Pu3VULm1EB0u/X9H6KHNgxMu59v71jH08tzTXZkP1zV4AS79f0fooc2DEy7m23tDqjSW69pMvf+QJMOv9itZHmQMjXs61GWtZ1BQvzbV3rWKeAMu8XzH1UdYbGEe2nGsnnrj/6KhcOc+1V95UN+u6uuAoXZ4OZr1f8WAs9VHWGxhHtpxrt909FI+PzldYzrX1PPEVZr1fcca71kf8GxhHOPu1G2v16kpzbeYVMFSKAOaxAOaxAOaxAOaxAOaxAOaxSgn2M03Txky+nfhuTLXZ6aq1Q8ZviXO0A4or5TRhS7FXZCUGm9raunvNhPk5J656PPvzw9oGF7PzpC+Ydw296jErjN/2nqQdcBoiWFVY5YHNMH596PScE9mvxMAHgftasz7PmZ06lD/9pL5g9hqZqq+6xPht6fe1A05DnPo/BXt8XJydmH3W2XUnGHvpkvHnrWGXatq1LLhw8udr32Os6qVJP2FVe9JDkp9nz04eSp1vn7TnexOnvcuyxicX5NfYe+P5F/7VvInqVSfrV2Hnr9EfifoQ8w7MPq49/ZIXJnRkbiU1LbWMsepd1+jjHv1a+qrW0IPTT/v2ixUJNrzvy9NZ4rLLt++4/NLEkc/U71up7U3+q3vN1TvaZp0dYFVTGj7UX0lriPl51uzkIet8+7grDsfumxTOHm8syK9xdefwr04x/z+I6k3X/pqxNyf8Nwlm3YHR0JTavZurx3RkbsWcll5GX7VlbDdj31hhXdUa2ls9feez551UcWCa0bc62faqI4x9NHZn85jDjG3uNF6J1pN7dI9zX2RVq5nxYltDzM+zZicPWefbtX8wFr9gbfZ4EyxzbB1jR7T3k8tUb9pUw9iDsw4nwaw7MNp8eoSxRq0jcyvmtPQy+qqJLzawjjEfWle1hm78gv690KtaxYHp3za0vp9gbP1FxqcXbIh8d8Kcp2PJ17dRq9LTfseqXmPGi20NMT/Pmp08ZJ1v14w3d964OHu8CZY5to0xXxrMV9XJvrrZBEvdwWb9X4UDDxuPu/e0jsytmNPSyxir3v5j9tCVzLqqNXSJ8Xbk/jEVBzYj9cHarxi/XrSWsd23XTjp78Yr8ewZqXP6nyvGL+khyc+zZicPWefbNf0ZxWYsyh5vgmUdy4Cx6b95Z3zUBEvdwWAwGIw/OE0/3al1ZN2KPi29jLHqjnHhi9Yz66rW0HsMsIHKeyRaL/mWqi79j/iqbTv0x1ei9qfGK3FQe5ux7uvfscCsIU5g1vl27W+Mxc78Q/Z4EyzrGAe28bJHZjITzLoDoz+fEWXsT1pH1q3o09LLGKvGz1x6so9ZV7WGbpqoP023Vt4j0XrJE9+5smXPFTWJ10564j/PX/Awm3pbgN1w8bZd104ZssCsIU5g1vl2bfILe2snhrLHGwva1uDATow99y8pMOsOjPrPu6H9la9XvZt1K/q09DLGqmzh2FrGrKtaQ6Nn/+j1V6Z8rmLB2PFZZ501W/9+e9X54760dJA1TLyBRW8797TrP0g/EtNDHMCs8+3aKxefes07LHu8saBtDd948+9uOhibNi6cArPuIFnnDz5/xf4qX9atGNOsZYxVWbP2DMtc1Rp6aPpp33xu9r4ReNEq7b8ltmtFe5PCsWf1X97+bEI86sVTY0W9aqEA5tinp/z22MGp4jdhD4em/qK4Vy0UwJz7Z834c+aL377Wqk3uLvJVC1RpYKNc/NCoSaUCmMcCmMcCmMcCmMcCmMcCmMcCmMf6H/5CQo706gnTAAAAAElFTkSuQmCC" /><!-- --></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references">
<div id="ref-gerard2018genotyping">
<p>Gerard, David, Luis Felipe Ventorim Ferrão, Antonio Augusto Franco Garcia, and Matthew Stephens. 2018. “Genotyping Polyploids from Messy Sequencing Data.” <em>Genetics</em> 210 (3). Genetics: 789–807. <a href="https://doi.org/10.1534/genetics.118.301468">https://doi.org/10.1534/genetics.118.301468</a>.</p>
</div>
<div id="ref-gerard2019priors">
<p>Gerard, David, and Luís Felipe Ventorim Ferrão. 2019. “Priors for Genotyping Polyploids.” <em>bioRxiv</em>. Cold Spring Harbor Laboratory. <a href="https://doi.org/10.1101/751784">https://doi.org/10.1101/751784</a>.</p>
</div>
<div id="ref-uitdewilligen2013next">
<p>Uitdewilligen, Anne-Marie A. AND D’hoop, Jan G. A. M. L. AND Wolters. 2013. “A Next-Generation Sequencing Method for Genotyping-by-Sequencing of Highly Heterozygous Autotetraploid Potato.” <em>PLOS ONE</em> 8 (5). Public Library of Science: 1–14. <a href="https://doi.org/10.1371/journal.pone.0062355">https://doi.org/10.1371/journal.pone.0062355</a>.</p>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
